### ISSUE TITLE
Bug report - Runtime write permissions for external SD Card

### ISSUE THREAD

**Issue by ghost** [CypherpunkArmory/UserLAnd#630]
# Describe the bug

Since Android 6.0.1 ext.-card files can't be written etc. regardless of whether the app _has_ access permission.

It has to prompt. More context below. Zzz...

# Steps to reproduce the behavior:

- Boot a Linux Terminal or app.
- Do `$ cd $HOME/host-rootfs/storage/emulated/<SD Card name>`.
- Write or `mkdir` and it will fail with `'permission denied'`.

# Device Information

Device: Samsung Galaxy J500M.
Android Version: Marshmallow 6.0.1.
UserLAnd Version: 2.3.4.
Whether the device is rooted: no.
Whether the device is running LineageOS: no.

# Additional context

## What I expected

A **_prompt_** for selecting the card to grant app's access at run-time once.

![](https://i.imgur.com/HTivJyb.png)

* [Universal way to write to external SD card on Android](https://stackoverflow.com/q/40068984#40126690)

**Comment by lewisdonofrio** [comment]
is this about UserLAnd cause this read liks your just having issues getting to the sd card?  I have this ticket https://github.com/CypherpunkArmory/UserLAnd/issues/591 that I'm guessing is what your trying to get to?

**Comment by hydroper** [comment]
@lewisdonofrio Yeah, didn't see it. Not all Android apps (including Quoda and UserLAnd) give SD Card clients the select prompt.

QuickEdit does as well as... ZArchiver.  [Universal way to write to external SD card on Android](https://stackoverflow.com/questions/40068984/universal-way-to-write-to-external-sd-card-on-android/40126690)

**Comment by hydroper** [comment]
@enforce-issue-templates

**Comment by hydroper** [comment]
Truly since Lollipop (Android 5.1) this SDCard restrictiom had been introduced.

I got back to 5.1 in this cell phone, rooted it with KingoRoot and it worked!  *_*

The ad-hoc solution is to root and, in some cases, use apps like SDCard Fix. I'm só happy, damn.

**Comment by slycooper724** [comment]
I can't write to my external sdcard either.

**Comment by corbinlc** [comment]
Our app has a prompt requesting sdcard access when you try to create a new filesystem.  Are you not seeing it and accepting it?  Once you have created a filesystem, `/sdcard` is where you will find your sdcards contents.

**Comment by corbinlc** [comment]
Specifically, prompt shows up when you create the first filesystem.

**Comment by XeGrox** [comment]
He meant the external sd card

**Comment by corbinlc** [comment]
That is what I mean to.  Should work for real or emulated sdcards.

**Comment by corbinlc** [comment]
So, @hydroper or others, can you try using our app from scratch again and tell us if you see the prompt when you go to create your first filesystem.  It should prompt you in a fashion very similar to what you have requested.

**Comment by hydroper** [comment]
> Our app has a prompt requesting sdcard access when you try to create a new filesystem.

Though prompts should occur everytime you re-open the UserLAnd app and _not_ once you create a filesystem. QuickEdit works same way here. I think it checks permissions, somehow...

For creating a filesystem in a external SD Card does one need to move the app there too?

**Comment by m-yosefpor** [comment]
I have the same problem. I can read from and write to internal sdcard located at "/host-rootfs/storage/emulated/0" , and also I can read from external storage located at "/host-rootfs/storage/XXXX-XXXX" (where X is a hexcharecter) but I can not write into that.
for example, the following command gives me a permission denied error:
touch /host-rootfs/storage/xxxx-xxxx

**Comment by m-yosefpor** [comment]
however I found a good technique here:

https://github.com/CypherpunkArmory/UserLAnd/issues/70#issuecomment-417438201

**Comment by bricolodu13** [comment]
Hello all,

I have two Android phones
One is running Android 6.0 and can access the SDCard from within UserLand Debian in root mode with ln -s /host-rootfs/sdcard /

The second one is running Android 8.1.0 and CANNOT access the SDCard from within UserLand Debian in root mode. The same command ln -s /host-rootfs/sdcard / return premission denied

I also tried under root permission (sudo su) to change the repertory rights with chmod -R 777 /host-rootfs and I got :
chmod : changing permissions of '/host-rootfs': Read-only file system
chmod : cannot read directory '/host-rootfs': Permission denied

While the same command work on Android 6.0

Please help as my Android phone 6.0 has issues and I would like to use my new phone.

NOTE that, I mostly just need read access
Best regards,
Jean

**Comment by bricolodu13** [comment]
I found a weird way to make it working.
I made a copy of my sdcard to the internal memory (I used File Commander app)
then, from the command line under root (sudo su) i moved the files to the home directory:
cd /sdcard
mv Music /home

Now I can play my Music but it is filling the internal memory !

**Comment by ytrezq** [comment]
@corbinlc used latest beta version from the store today, and I got no prompt at all.

**Comment by cld71** [comment]
Running UserLAnd 2.7.0 on a Samsung Galaxy S7.
Not able to write to the SD card directly, but am able to write to /storage/sdcard with no problem.
Which is the same as SDCard/Android/data/tech.ula/files/storage file.

The only other thing is that unable to execute any shell files from /storage/sdcard, or /storage/internal .

**Comment by corbinlc** [comment]
Closing as part of a repository cleanup. If this issue is still relevant, feel free to reopen it or file a fresh one.
